race-buffer/model/RaceBuffer.tla

```diff
@@ -6,111 +6,93 @@ CONSTANTS
 
   NumWrites  (* Number of writes to execute *)
 
-ASSUME Assumption == BufCapacity \in (Nat \ 0..2) (* Buffer capacity is at least 3 *)
+ASSUME Assumption == BufCapacity \in Nat\*(Nat \ 0..2) (* Buffer capacity is at least 3 *)
 
 -----------------------------------------------------------------------------
 
 VARIABLES 
   wpc, (* Writer's program counter *)
   buf, (* Backing storage *)
-  wcurs, (* Write cursor - Equal to number of entries already written *)
-
-  writes, (* Number of entries written (used to limit state space) *)
+  wcurs, (* Write cursor - Points to next write position *)
+  owcurs, (* Overwrite cursor - Marks where space is allocated for the writer to write freely *)
 
   rpc, (* Reader's program counter *)
   rbuf, (* Buffer of entries already read *)
   bufSnap, (* Intermediate buffer used to store a snapshot of entries being read *)
-  rcurs, (* Read cursor - Equal to number of entries already read*)
-  wcursSnap (* Value of wcurs observed at beginning of read *)
+  rcurs, (* Read cursor - Points to next read position *)
+  wcursSnap, (* Value of write cursor observed at beginning of read *)
+  storedPrefix, (* Prefix to be written to read buffer once suffix is read *)
+  processingIndex (* Index of next entry in bufSnap to process *)
 
 -----------------------------------------------------------------------------
 
 (* Writer helpers *)
 
 (* Construct an entry structure - Index field is only used to check *)
-(* ordering in the read buffer of non-nil entries. *)
+(* ordering in the reader's output buffer *)
 Entry(i, t) == [index |-> i, type |-> t]
 
-(* Get index in backing storage based on entry index *)
+(* Get index in backing storage based on given cursor *)
 BufIndex(i) == i % BufCapacity
 
-(* Get entry in storage position of given index *)
+(* Get entry in storage position of given cursor *)
 Read(i) == buf[BufIndex(i)]
 
 (* Write a single entry of given type at given index *)
 Write(i, t) == buf' = (BufIndex(i) :> Entry(i,t)) @@ buf
 
-(* Write the appropriately typed non-nil entry at the given index *)
+(* Either write a single entry or the prefix of a double entry *)
+WriteNewEntry(i) == 
+  \/ Write(i, "SINGLE")
+  \/ Write(i, "PREFIX")
+
+(* Write the appropriately typed entry at the given index *)
 WriteEntry(i) == 
-  /\ IF Read(i - 1).type = "PREFIX"
-     THEN Write(i, "SUFFIX")
-     ELSE \/ Write(i, "SINGLE")
-          \/ Write(i, "PREFIX")
-  /\ writes' = writes + 1
+  IF i = 0 
+  THEN WriteNewEntry(i) 
+  ELSE IF Read(i - 1).type = "PREFIX"
+       THEN Write(i, "SUFFIX")
+       ELSE WriteNewEntry(i)
+
+(* Overwrite old entry by incrementing overwrite cursor *)
+OverwriteOldEntry == 
+    IF owcurs < BufCapacity 
+    THEN owcurs' = owcurs + 1
+    ELSE IF Read(owcurs).type = "PREFIX" 
+         THEN owcurs' = owcurs + 2
+         ELSE owcurs' = owcurs + 1
```

Comment:
I think this is an actual read dependency

Reply:
The writer is the only one that mutates writer state. It wouldn't make a difference if the reader were to read things before or after the `Read(owcurs).type = "PREFIX" ` check.

Reply:
But, again, maybe ok in this case, since this is single writer?

Reply:
> But, again, maybe ok in this case, since this is single writer?

I believe so

Reply:
Yes, thinking about this more, the atomic steps that can be interspersed in any order are single writes to writer memory by the writer and single reads from writer memory by the reader.